Windows: Add regex filtering to dumpfiles

brandon-barnacle opened the pull request:
In volatility2, the windows.dumpfiles plugin allows you to filter the dumped files using a regular expression. This PR adds the same functionality to volatility3.

The regular expression is passed in using --regex=REGEX and all files matching REGEX will be dumped. The --ignore-case flag can be passed to make the search case-insensitive. The search is case-sensitive by default. The matching volatility2 functionality can be found here:
https://github.com/volatilityfoundation/volatility/blob/a438e768194a9e05eb4d9ee9338b881c0fa25937/volatility/plugins/dumpfiles.py#L844

Manual testing was performed on windows memory images across different windows versions to verify the expected output.

brandon-barnacle committed:
Windows: Add regex filtering to dumpfiles

In volatility2, the windows.dumpfiles plugin allows you to filter the dumped files
using a regular expression. This PR adds the same functionality to volatility3.

The regular expression is passed in using --regex=REGEX and all files matching
REGEX will be dumped. The --ignore-case flag can be passed to make the search
case-insensitive. The search is case-sensitive by default. The matching
volatility2 functionality can be found here:
https://github.com/volatilityfoundation/volatility/blob/a438e768194a9e05eb4d9ee
9338b881c0fa25937/volatility/plugins/dumpfiles.py#L844

Manual testing was performed on windows memory images across different windows
versions to verify the expected output.

ikelos commented on volatility3/framework/plugins/windows/dumpfiles.py:
```diff
@@ -243,6 +258,14 @@ def _generator(self, procs: List, offsets: List):
                         if obj_type == "File":
                             file_obj = entry.Body.cast("_FILE_OBJECT")
+
+                            if self.config["regex"]:
```
Presumably this could/should just check whether file_re is set?  Going back to the config allows the (admittedly, extremely unlikely) possibility that something changes the config between when you checked it and when you check it again.

brandon-barnacle replied:
Yes this makes sense! I will change in both spots.

ikelos commented on volatility3/framework/plugins/windows/dumpfiles.py:
```diff
@@ -272,6 +295,13 @@ def _generator(self, procs: List, offsets: List):
                         if not file_obj.is_valid():
                             continue
 
+                        if self.config["regex"]:
```
Same as above.

ikelos commented on volatility3/framework/plugins/windows/dumpfiles.py:
```diff
@@ -53,6 +54,15 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
                 description="Dump a single _FILE_OBJECT at this physical address",
                 optional=True,
             ),
+            requirements.StringRequirement(
```
So adding filtering options like this is great, but also needs some consideration.  Is it consistent with other filter options?  `regex` isn't that descriptive a name, I'd prefer `filter` or `match` or something like that, and then let the `--help` describe that it's a regular expression.  Most people will know what regexs are, but not everybody, and it doesn't really describe the purpose of it.

ikelos replied:
Consideration also needs to be used to determine how this option interacts with others?  If people provide `virtaddr` and `regex`, which wins? (or are they both applied?  Does that make sense?)  What about `pid`?  It should be clear and/or obvious which options will interact with each other and which cancel others out.

brandon-barnacle replied:
The flag name change definitely makes sense and `--filter` makes a lot of sense, so I think changing to that will be good.

In terms of how it will interact with the other options, I think it makes sense that it cannot interact with the flags that take in an address, but it can work with the `pid` flag. My newest commit addresses this.

ikelos commented on volatility3/framework/plugins/windows/dumpfiles.py:
```diff
@@ -208,6 +218,11 @@ def process_file_object(
         kernel = self.context.modules[self.config["kernel"]]
+        if self.config["regex"]:
+            if self.config["ignore-case"]:
+                file_re = re.compile(self.config["regex"], re.I)
```
Also, `file_re` needs to be initialized.

ikelos reviewed (commented):
Generally the coding style looks good, and the changes are minimal, but adding options to a plugin always needs some thinking about, so there's just some questions about the precedence of the command line options and how that would work.

brandon-barnacle committed:
PR comment changes

Change the --regex flag to --filter. Add a check so that --filter
cannot be used with --physaddr or --virtaddr. Change
self.config["filter"] check to check if file_re has been set.

ikelos reviewed (commented):
Thanks, this looks much better.  Please could you make sure it's formatted by black and then we can get it merged...  5:)

ikelos commented:
So the rest of the codebase was still formatted against 23.12.0 rather than 24.1.0 (which is what github now uses).  The issues it's finding are from the original code, so this should merge in just fine and I'll fix it up manually if not...
